File: LinkedList/Cpp/Problem10.cpp
/*
Delete head from a DLL.
*/

#include <bits/stdc++.h>
using namespace std;

struct DLL
{
  int data;
  DLL *next;
  DLL *prev;
  DLL(int x)
  {
    data = x;
    next = NULL;
    prev = NULL;
  }
};

DLL *deleteHead(DLL *head)
{
  if (head == NULL)
    return NULL;
  else if (head->next == NULL)
  {
    delete head;
    return NULL;
  }
  else
  {
    DLL *curr = head->next;
    delete head;
    // Deleting head in this function does not delete head globally.
    curr->prev = NULL;
    return curr;
  }
}

DLL *insertStart(DLL *head, int x)
{
  DLL *temp = new DLL(x);
  if (head == NULL)
  {
    return temp;
  }
  temp->next = head;
  head->prev = temp;
  return temp;
}

void printList(DLL *head)
{
  DLL *curr = head;
  while (curr != NULL)
  {
    cout << curr->data << " ";
    curr = curr->next;
  }
  cout << endl;
}

int main()
{
  DLL *head = NULL;
  for (int i = 10; i > 0; i--)
    head = insertStart(head, i);
  printList(head);
  head = deleteHead(head);
  printList(head);
  return 0;
}

File: LinkedList/Java/Problem10.java
/*
Delete head of DLL.
*/

import java.io.*;

class DLL {
  int data;
  DLL next;
  DLL prev;

  DLL(int x) {
    data = x;
    next = null;
    prev = null;
  }
}

class GFG {
  public static DLL deleteHead(DLL head) {
    if (head == null || head.next == null)
      return null;
    else {
      head = head.next;
      head.prev = null;
      return head;
    }
  }

  public static DLL insertEnd(DLL head, int x) {
    DLL temp = new DLL(x);
    if (head == null) {
      return temp;
    }
    DLL curr = head;
    while (curr.next != null) {
      curr = curr.next;
    }
    curr.next = temp;
    temp.prev = curr;
    return head;
  }

  public static void printList(DLL node) {
    while (node != null) {
      System.out.print(node.data + " ");
      node = node.next;
    }
    System.out.println("");
  }

  public static void main(String[] args) {
    DLL head = null;
    for (int i = 10; i <= 100; i += 10)
      head = insertEnd(head, i);
    printList(head);
    head = deleteHead(head);
    printList(head);
  }
}
